### PR TITLE
storage: fix file deletion check to match POSIX

### PR DESCRIPTION
**Pull request by kofemann**
according to POSIX[1], we can remove file in a directory
if we can write into directory. Optionally, we can remove
file writable if can write to it. But this is file-system
implementation specific.

Fix the Storage::Common#isDeletable() to match POSIX

While our production system affected by this issue, please consider this fix for stable branches (8.2) as well.

[1] http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_02

Signed-off-by: Tigran Mkrtchyan tigran.mkrtchyan@desy.de

### CODE REVIEW

**Comment by kofemann** [comment]
MIT licensed

File: lib/private/files/storage/common.php [hunk 1]
```diff
@@ -133,7 +133,9 @@ public function isDeletable($path) {
 			return false;
 		}
 		$parent = dirname($path);
-		return $this->isUpdatable($parent) && $this->isUpdatable($path);
+		// according to POSIX, we can delete file is directory allowes that
```
Comment (DeepDiver1975) [inline lib/private/files/storage/common.php]: if

**Comment by DeepDiver1975** [comment]
@icewind1991

File: lib/private/files/storage/common.php (hunk 1 above)
Comment (kofemann) [inline lib/private/files/storage/common.php]: thanks. fixed

**Comment by icewind1991** [comment]
:+1: looks good

**Comment by Xenopathic** [comment]
The document you reference applies only to directories with the sticky bit set, while normal directories without it are affected by slightly different rules, namely deletion is possible if the directory is writeable (no mention of the file itself). However, deleting a (writeable) file in a directory that isn't writeable will give a permission denied error.

So :-1:

**Comment by kofemann** [comment]
If there is no sticky bit, then you don't check for files owner ship.

**Comment by kofemann** [comment]
as reference :
https://github.com/torvalds/linux/blob/master/fs/namei.c#L2530

**Comment by Xenopathic** [comment]
Correct, but your change made it so that write access on the directory is optional (by changing 'and' to 'or'), which is wrong.

**Comment by kofemann** [comment]
The write test on the file itself.

**Comment by kofemann** [comment]
any update here?

**Comment by Xenopathic** [comment]
No need for any update, this change is wrong for the reasons I stated above. Think about a (writable) file in a (read-only) directory: this change would make it 'deletable', while it actually isn't.

**Comment by kofemann** [comment]
Yes I am fully agree that remove must check directory permissions only. Now checking looking at master again I see that it has the correct behavior fixed by commit ab79caf2 

Can you get it into 8.2 branch (and close this pull request)

**Comment by kofemann** [comment]
here is the commint which introduces the problem: 2f22e67570bd7b3e99d0f01b59c0e1808a8b5189

**Comment by MorrisJobke** [comment]
> Can you get it into 8.2 branch (and close this pull request)

cc @Xenopathic @icewind1991

**Comment by Xenopathic** [comment]
That commit you mentioned is in 8.2.1

**Comment by kofemann** [comment]
exactly,  2f22e67 in 8.2.1 and this is wrong behavior.

**Comment by Xenopathic** [comment]
I was more talking about ab79caf29bf4173e15f9dad00a206c6bb1f3999c, which you mentioned above

**Pull request closed without merging**